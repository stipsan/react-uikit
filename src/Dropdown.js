import cx from 'classnames'
import { Component, PropTypes, cloneElement } from 'react'

const links = {}

const subscribe = (namespace, listener) => {
  if (!links[namespace]) {
    links[namespace] = []
  }

  let isSubscribed = true

  links[namespace].push(listener)

  return function unsubscribe() {
    if (!isSubscribed) {
      return
    }

    isSubscribed = false

    const index = links[namespace].indexOf(listener)
    links[namespace].splice(index, 1)
  }
}
const dispatch = (namespace) => {
  const listeners = links[namespace]
  for (let i = 0; i < listeners.length; i++) { // eslint-disable-line no-plusplus
    const listener = listeners[i]
    listener()
  }
}

export default class Dropdown extends Component {
  static propTypes = {
    children: PropTypes.oneOfType([
      PropTypes.func,
      PropTypes.arrayOf(PropTypes.element),
    ]).isRequired,
    delay: PropTypes.number.isRequired,
    mode: PropTypes.oneOf(['hover', 'click']).isRequired,
    remainTime: PropTypes.number.isRequired,
    className: PropTypes.string,
    component: PropTypes.node,
    link: PropTypes.string,
  }

  static defaultProps = {
    mode: 'hover',
    remainTime: 800,
    delay: 0,
    component: 'div',
    link: '',
    className: 'uk-inline',
  }

  state = {
    isOpen: false,
  }

  componentDidMount() {
    if (this.props.link) {
      // listen to linked components
      this.unsubscribe = subscribe(this.props.link, () => {
        if (this.state.isOpen) {
          this.setState({ isOpen: false })
        }
      })
    }
  }

  componentWillUnmount() {
    if (this.leaveTimeout) {
      clearTimeout(this.leaveTimeout)
    }

    if (this.unsubscribe) {
      this.unsubscribe()
      this.unsubscribe = false
    }
  }

  handleMouseEnter = () => {
    if (this.leaveTimeout) {
      clearTimeout(this.leaveTimeout)
    }
    if (this.props.delay) {
      setTimeout(() => {
        this.setState({ isOpen: true })
      }, this.props.delay)
    } else {
      if (this.props.link) {
        dispatch(this.props.link)
      }
      this.setState({ isOpen: true })
    }

    if (this.props.mode === 'hover') {
      this.setState({ isOpen: true })
    }
  }

  handleMouseLeave = () => {
    if (this.props.remainTime) {
      this.leaveTimeout = setTimeout(() => {
        this.setState({ isOpen: false })
      }, this.props.remainTime)
    } else {
      this.setState({ isOpen: false })
    }
  }

  handleClick = () => {
    if (this.props.mode === 'click') {
      this.setState({ isOpen: !this.state.isOpen })
    }

    if (this.props.mode === 'hover') {
      this.setState({ isOpen: false })
    }
  }

  render() {
    const { handleMouseEnter, handleMouseLeave, handleClick } = this
    const { mode, component, children, className } = this.props
    const { isOpen } = this.state

    const eventHandlers = {
      onClick: handleClick,
      onMouseEnter: mode === 'hover' && handleMouseEnter,
      onMouseLeave: mode === 'hover' && handleMouseLeave,
    }

    if (typeof children === 'function') {
      return children({ isOpen, ...eventHandlers })
    }

    if (!Array.isArray(children) || children.length !== 2) {
      throw new Error('Children must be passed as array and must have two components.')
    }
    const [target, body] = children

    // construct target Element
    const targetElement = cloneElement(target, {
      ...target.props,
      ...eventHandlers,
      className: cx(target.props.className, {
        'uk-open': isOpen,
      }),
    })

    // construct body Element
    const bodyElement = cloneElement(body, {
      ...body.props,
      ...eventHandlers,
      className: cx(body.props.className, {
        'uk-open': isOpen,
      }),
    })

    const WrapperComponent = component

    return (
      <WrapperComponent className={className}>
        {targetElement}
        {bodyElement}
      </WrapperComponent>
    )
  }
}
