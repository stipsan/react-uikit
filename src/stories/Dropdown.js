import { storiesOf } from '@kadira/storybook'
import cx from 'classnames'
import { PropTypes, Component } from 'react'
import Dropdown from '../Dropdown'
import Button from '../Button'

// This is to work around: https://github.com/kadirahq/react-storybook-addon-info/issues/26#issuecomment-229029177
Dropdown.displayName = 'Dropdown'

const Nav = () => (
  <ul className="uk-nav uk-nav-dropdown">
    <li><a className="uk-warning" href="#">Item</a></li>
    <li><a href="#">Another item</a></li>
    <li className="uk-nav-header">Header</li>
    <li><a href="#">Item</a></li>
    <li><a href="#">Another item</a></li>
    <li className="uk-nav-divider" />
    <li><a href="#">Separated item</a></li>
  </ul>
)

const Menu = () => <Nav />
Menu.displayName = 'Menu'

const LoremIpsum = `Lorem ipsum dolor sit amet, consectetur adipisicing elit,
    sed do eiusmod tempor incididunt ut labore et dolore magna aliqua.`

class MyDropdown extends Component {
  static propTypes = {
    className: PropTypes.string,
    onMouseEnter: PropTypes.func,
    onMouseLeave: PropTypes.func,
  }

  handleMouseLeave = () => {
    this.props.onMouseLeave()
  }

  handleMouseEnter = () => {
    this.props.onMouseEnter()
  }

  render() {
    const { ...props } = this.props
    const { handleMouseEnter, handleMouseLeave } = this
    return (
      <div
        {...props}
        className={cx('uk-dropdown uk-dropdown-bottom-left', props.className)}
        onMouseEnter={handleMouseEnter}
        onMouseLeave={handleMouseLeave}
      >
        {LoremIpsum}
      </div>
    )
  }
}

storiesOf('Dropdown', module)
    .addWithInfo('Basic Usage', '', () => (
      <div className="uk-margin-bottom">
        <Dropdown mode="hover">
          <Button>Hover</Button>
          <MyDropdown />
        </Dropdown>
        &nbsp;
        <Dropdown mode="click">
          <Button primary>Click</Button>
          <MyDropdown />
        </Dropdown>
      </div>
), { header: false, inline: true, propTables: [Dropdown] })

  .addWithInfo('Nav in dropdown', '', () => (
    <div className="uk-margin-bottom">
      <Dropdown>
        <Button>Hover</Button>
        <div className="uk-dropdown uk-dropdown-bottom-left">
          <Nav />
        </div>
      </Dropdown>
    </div>
  ), { header: false, inline: true, propTables: [Dropdown] })

  .addWithInfo('Advanced', '', () => (
    <div className="uk-margin-bottom">
      <nav className="uk-navbar-container">
        <ul className="uk-navbar-nav">
          <li className="uk-active"><a>Active</a></li>
          <li><a href="">Item</a></li>
          <Dropdown className="" component="li" >
            <a href="#">Parent</a>
            <div className="uk-navbar-dropdown">
              <ul className="uk-nav uk-navbar-dropdown-nav">
                <li className="uk-active"><a href="#">Active</a></li>
                <li><a href="#">Item</a></li>
                <li className="uk-nav-header">Header</li>
                <li><a href="#">Item</a></li>
                <li><a href="#">Item</a></li>
                <li className="uk-nav-divider" />
                <li><a href="#">Item</a></li>
              </ul>
            </div>
          </Dropdown>
          <li><a href="">Item</a></li>
        </ul>
      </nav>
    </div>
  ), { header: false, inline: true })
