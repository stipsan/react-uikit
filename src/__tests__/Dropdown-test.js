import renderer from 'react-test-renderer'

import Dropdown from '../Dropdown'

jest.useFakeTimers()

describe('Dropdown', () => {
  it('renders correctly', () => {
    expect(renderer.create(
      <Dropdown>
        <button className="uk-button">Hover <i className="uk-icon-caret-down" /></button>
      </Dropdown>
    ).toJSON()).toMatchSnapshot()

    expect(renderer.create(
      <Dropdown className="uk-parent" component="li" >
        <a>Parent <i className="uk-icon-caret-down" /></a>
      </Dropdown>
    ).toJSON()).toMatchSnapshot()
  })

  it('changes the class when hovered', () => {
    let component = renderer.create(
      <Dropdown>
        <button className="uk-button">Hover <i className="uk-icon-caret-down" /></button>
      </Dropdown>
    )
    let tree = component.toJSON()
    expect(tree).toMatchSnapshot()

    // manually trigger the callback
    tree.props.onMouseEnter()
    expect(component.toJSON()).toMatchSnapshot()

    // manually trigger the callback
    tree.props.onMouseLeave()
    expect(component.toJSON()).toMatchSnapshot()

    // Fast forward timers, class should now be gone
    jest.runAllTimers()
    expect(component.toJSON()).toMatchSnapshot()

    let instance = component.getInstance()
    instance.handleMouseEnter()
    expect(instance.state.isOpen).toBeTruthy()

    component = renderer.create(
      <Dropdown delay={400} mode="hover" remainTime={0}>
        <button className="uk-button">Hover <i className="uk-icon-caret-down" /></button>
      </Dropdown>
    )
    instance = component.getInstance()
    instance.handleMouseEnter()
    expect(instance.state.isOpen).toBeTruthy()

    jest.runAllTimers()
    tree = component.toJSON()
    expect(tree).toMatchSnapshot()

    instance.handleMouseLeave()
    jest.runAllTimers()
    expect(tree).toMatchSnapshot()
    expect(instance.state.isOpen).toBe(false)
  })

  it('changes the class when clicked', () => {
    const component = renderer.create(
      <Dropdown mode="click">
        <button className="uk-button">Click <i className="uk-icon-caret-down" /></button>
      </Dropdown>
    )
    let tree = component.toJSON()
    expect(tree).toMatchSnapshot()

    // manually trigger the callback
    tree.props.onClick()
    // re-rendering
    tree = component.toJSON()
    expect(tree).toMatchSnapshot()

    // manually trigger the callback
    tree.props.onClick()
    // re-rendering
    tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('closes sibling Dropdown on open', () => {
    const firstComponent = renderer.create(
      <Dropdown link="menu">
        <button className="uk-button">First Instance <i className="uk-icon-caret-down" /></button>
      </Dropdown>
    )
    let first = firstComponent.toJSON()
    expect(first).toMatchSnapshot()

    // manually trigger the callback
    first.props.onMouseEnter()

    first = firstComponent.toJSON()
    expect(first).toMatchSnapshot()
    expect(firstComponent.getInstance().state.isOpen).toBe(true)

    const secondComponent = renderer.create(
      <Dropdown link="menu">
        <button className="uk-button">Second Instance <i className="uk-icon-caret-down" /></button>
      </Dropdown>
    )
    let second = secondComponent.toJSON()
    expect(second).toMatchSnapshot()

    // manually trigger the callback
    second.props.onMouseEnter()

    // Verify that it is open
    second = secondComponent.toJSON()
    expect(second).toMatchSnapshot('should be open')
    expect(secondComponent.getInstance().state.isOpen).toBe(true)
    // Verify that the first component is closed
    first = firstComponent.toJSON()
    expect(first).toMatchSnapshot('should be closed')
    expect(firstComponent.getInstance().state.isOpen).toBe(false)

    // Verify there is a subscription
    const secondInstance = secondComponent.getInstance()
    expect({}.hasOwnProperty.call(secondInstance, 'unsubscribe')).toBe(true)
    const unsubscribe = secondInstance.unsubscribe

    // Start leave timeout before unmount
    second.props.onMouseLeave()

    // Trigger unmount
    secondComponent.update(<span />)
    // Should be no error from calling this again
    unsubscribe()
  })

  it('closes when clicked when in hover mode', () => {
    const component = renderer.create(
      <Dropdown mode="hover">
        <button className="uk-button">Hover <i className="uk-icon-caret-down" /></button>
      </Dropdown>
    )
    let tree = component.toJSON()

    // clicking the dropdown should not open it
    tree.props.onClick()
    tree = component.toJSON()
    expect(component.getInstance().state.isOpen).toBe(false)

    // open the dropdown
    tree.props.onMouseEnter()
    tree = component.toJSON()
    expect(tree).toMatchSnapshot()
    expect(component.getInstance().state.isOpen).toBe(true)

    // close the dropdown by click
    tree.props.onClick()
    tree = component.toJSON()
    expect(tree).toMatchSnapshot()
    expect(component.getInstance().state.isOpen).toBe(false)

    // clicking the dropdown again shouldn't open it
    tree.props.onClick()
    tree = component.toJSON()
    expect(tree).toMatchSnapshot()
    expect(component.getInstance().state.isOpen).toBe(false)
  })
})
