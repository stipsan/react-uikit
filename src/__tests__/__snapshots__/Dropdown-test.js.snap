// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`changes the class when clicked 1`] = `
<div
  className="uk-inline"
>
  <button
    className="uk-button uk-button-default"
    onClick={[Function]}
    onMouseEnter={false}
    onMouseLeave={false}
    type="button"
  >
    Click
  </button>
  <div
    className="uk-dropdown uk-dropdown-bottom-left"
    onClick={[Function]}
    onMouseEnter={false}
    onMouseLeave={false}
  >
    Lorem ipsum dolor sit amet, consectetur adipisicing elit
  </div>
</div>
`;

exports[`changes the class when clicked 2`] = `
<div
  className="uk-inline"
>
  <button
    className="uk-button uk-open uk-button-default"
    onClick={[Function]}
    onMouseEnter={false}
    onMouseLeave={false}
    type="button"
  >
    Click
  </button>
  <div
    className="uk-dropdown uk-dropdown-bottom-left uk-open"
    onClick={[Function]}
    onMouseEnter={false}
    onMouseLeave={false}
  >
    Lorem ipsum dolor sit amet, consectetur adipisicing elit
  </div>
</div>
`;

exports[`changes the class when clicked 3`] = `
<div
  className="uk-inline"
>
  <button
    className="uk-button uk-button-default"
    onClick={[Function]}
    onMouseEnter={false}
    onMouseLeave={false}
    type="button"
  >
    Click
  </button>
  <div
    className="uk-dropdown uk-dropdown-bottom-left"
    onClick={[Function]}
    onMouseEnter={false}
    onMouseLeave={false}
  >
    Lorem ipsum dolor sit amet, consectetur adipisicing elit
  </div>
</div>
`;

exports[`changes the class when hovered 1`] = `
<div
  className="uk-inline"
>
  <button
    className="uk-button uk-button-default"
    onClick={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    type="button"
  >
    Hover
  </button>
  <div
    className="uk-dropdown uk-dropdown-bottom-left"
    onClick={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    Lorem ipsum dolor sit amet, consectetur adipisicing elit
  </div>
</div>
`;

exports[`changes the class when hovered 2`] = `
<div
  className="uk-inline"
>
  <button
    className="uk-button uk-open uk-button-default"
    onClick={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    type="button"
  >
    Hover
  </button>
  <div
    className="uk-dropdown uk-dropdown-bottom-left uk-open"
    onClick={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    Lorem ipsum dolor sit amet, consectetur adipisicing elit
  </div>
</div>
`;

exports[`changes the class when hovered 3`] = `
<div
  className="uk-inline"
>
  <button
    className="uk-button uk-open uk-button-default"
    onClick={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    type="button"
  >
    Hover
  </button>
  <div
    className="uk-dropdown uk-dropdown-bottom-left uk-open"
    onClick={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    Lorem ipsum dolor sit amet, consectetur adipisicing elit
  </div>
</div>
`;

exports[`changes the class when hovered 4`] = `
<div
  className="uk-inline"
>
  <button
    className="uk-button uk-button-default"
    onClick={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    type="button"
  >
    Hover
  </button>
  <div
    className="uk-dropdown uk-dropdown-bottom-left"
    onClick={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    Lorem ipsum dolor sit amet, consectetur adipisicing elit
  </div>
</div>
`;

exports[`changes the class when hovered 5`] = `
<div
  className="uk-inline"
>
  <button
    className="uk-button uk-open uk-button-default"
    onClick={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    type="button"
  >
    Hover
  </button>
  <div
    className="uk-dropdown uk-dropdown-bottom-left uk-open"
    onClick={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    Lorem ipsum dolor sit amet, consectetur adipisicing elit
  </div>
</div>
`;

exports[`changes the class when hovered 6`] = `
<div
  className="uk-inline"
>
  <button
    className="uk-button uk-open uk-button-default"
    onClick={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    type="button"
  >
    Hover
  </button>
  <div
    className="uk-dropdown uk-dropdown-bottom-left uk-open"
    onClick={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    Lorem ipsum dolor sit amet, consectetur adipisicing elit
  </div>
</div>
`;

exports[`closes sibling Dropdown on open 1`] = `
<div
  className="uk-inline"
>
  <button
    className="uk-button uk-button-default"
    onClick={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    type="button"
  >
    Hover
  </button>
  <div
    className="uk-dropdown uk-dropdown-bottom-left"
    onClick={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    Lorem ipsum dolor sit amet, consectetur adipisicing elit
  </div>
</div>
`;

exports[`closes sibling Dropdown on open 2`] = `
<div
  className="uk-inline"
>
  <button
    className="uk-button uk-open uk-button-default"
    onClick={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    type="button"
  >
    Hover
  </button>
  <div
    className="uk-dropdown uk-dropdown-bottom-left uk-open"
    onClick={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    Lorem ipsum dolor sit amet, consectetur adipisicing elit
  </div>
</div>
`;

exports[`closes sibling Dropdown on open 3`] = `
<div
  className="uk-inline"
>
  <button
    className="uk-button uk-button-default"
    onClick={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    type="button"
  >
    Hover
  </button>
  <div
    className="uk-dropdown uk-dropdown-bottom-left"
    onClick={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    Lorem ipsum dolor sit amet, consectetur adipisicing elit
  </div>
</div>
`;

exports[`closes when clicked when in hover mode 1`] = `
<div
  className="uk-inline"
>
  <button
    className="uk-button uk-open uk-button-default"
    onClick={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    type="button"
  >
    Hover
  </button>
  <div
    className="uk-dropdown uk-dropdown-bottom-left uk-open"
    onClick={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    Lorem ipsum dolor sit amet, consectetur adipisicing elit
  </div>
</div>
`;

exports[`closes when clicked when in hover mode 2`] = `
<div
  className="uk-inline"
>
  <button
    className="uk-button uk-button-default"
    onClick={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    type="button"
  >
    Hover
  </button>
  <div
    className="uk-dropdown uk-dropdown-bottom-left"
    onClick={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    Lorem ipsum dolor sit amet, consectetur adipisicing elit
  </div>
</div>
`;

exports[`closes when clicked when in hover mode 3`] = `
<div
  className="uk-inline"
>
  <button
    className="uk-button uk-button-default"
    onClick={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    type="button"
  >
    Hover
  </button>
  <div
    className="uk-dropdown uk-dropdown-bottom-left"
    onClick={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    Lorem ipsum dolor sit amet, consectetur adipisicing elit
  </div>
</div>
`;

exports[`it should render correctly when children are passed as function 1`] = `
<div
  onClick={[Function]}
>
  <button
    className=""
  >
    Target
  </button>
  <div
    className="uk-navbar-dropdown"
  >
    Dropdown content here
  </div>
</div>
`;

exports[`it should throw Invalid proptytype error on invalid children 1`] = `
"Warning: Failed prop type: Invalid prop \`children\` supplied to \`Dropdown\`.
    in Dropdown"
`;

exports[`renders correctly 1`] = `
<div
  className="uk-inline"
>
  <button
    className="uk-button uk-button-default"
    onClick={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    type="button"
  >
    Hover
  </button>
  <div
    className="uk-dropdown uk-dropdown-bottom-left"
    onClick={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    Lorem ipsum dolor sit amet, consectetur adipisicing elit
  </div>
</div>
`;

exports[`renders correctly 2`] = `
<li
  className="uk-inline"
>
  <button
    className="uk-button uk-button-link"
    onClick={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    type="button"
  >
    Parent
  </button>
  <div
    className="uk-navbar-dropdown"
    onClick={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    <ul
      className="uk-nav uk-navbar-dropdown-nav"
    >
      <li
        className="uk-active"
      >
        Active
      </li>
      <li>
        Item
      </li>
    </ul>
  </div>
</li>
`;

exports[`should be closed 1`] = `
<div
  className="uk-inline"
>
  <button
    className="uk-button uk-button-default"
    onClick={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    type="button"
  >
    Hover
  </button>
  <div
    className="uk-dropdown uk-dropdown-bottom-left"
    onClick={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    Lorem ipsum dolor sit amet, consectetur adipisicing elit
  </div>
</div>
`;

exports[`should be open 1`] = `
<div
  className="uk-inline"
>
  <button
    className="uk-button uk-open uk-button-default"
    onClick={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    type="button"
  >
    Hover
  </button>
  <div
    className="uk-dropdown uk-dropdown-bottom-left uk-open"
    onClick={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    Lorem ipsum dolor sit amet, consectetur adipisicing elit
  </div>
</div>
`;
